ssl/ssl_lib.cc

```diff
@@ -1757,16 +1757,25 @@ int SSL_has_pending(const SSL *ssl) {
 }
 
 int SSL_CTX_check_private_key(const SSL_CTX *ctx) {
-  return ssl_cert_check_private_key(ctx->cert.get(),
-                                    ctx->cert->privatekey.get());
+  if (!ssl_cert_check_cert_private_keys_usage(ctx->cert.get())) {
+    return 0;
+  }
+  return ssl_cert_check_private_key(
+      ctx->cert.get(),
+      ctx->cert->cert_private_keys[ctx->cert->cert_private_key_idx]
+          .privatekey.get());
 }
 
 int SSL_check_private_key(const SSL *ssl) {
-  if (!ssl->config) {
```

Comment:
* These were probably left over from OpenSSL that hadn't been refactored. OpenSSL code has a really bad habit of not doing proper null checks. It might be worth cleaning these up all at once in another SIM/PR.
* The behavior is well documented in `SSL_set_shutdown` so I believe its intentional. `SSL_CTX_set_quiet_shutdown` is encouraged instead.